### PR TITLE
Fix for #9524: fix recursive included decorator style propagation

### PR DESCRIPTION
**Pull request by level420**
Replaced the faulty recursive decorator include algorithm.

See #9524.

### CODE REVIEW

**Review by cboulanger — DISMISSED** [review]

**Review by hkollmann — DISMISSED** [review]

File: framework/source/class/qx/theme/manager/Decoration.js [hunk 1]
```diff
@@ -214,54 +214,28 @@ qx.Class.define("qx.theme.manager.Decoration",
...
-        if (entry.decorator || included.decorator) {
-          result.decorator = qx.lang.Object.clone(entry.decorator || included.decorator);
+      // handle recursive decorator includes
+      var recurseEntryInclude = function(currentEntry) {
+        // follow the include chain to the topmost decorator entry
+        if(currentEntry.include && theme.decorations[currentEntry.include]) {
+          recurseEntryInclude(theme.decorations[currentEntry.include]);
+        }
+        
+        // if the docorator is not set yet, clone the topmost
```
Comment (derrell) [inline framework/source/class/qx/theme/manager/Decoration.js]: if the *decorator* is not set...

**Review by derrell — CHANGES_REQUESTED** [review]
Very minor. Just a typo. Sorry.

**Review by oetiker — DISMISSED** [review]

**Comment by Adolfrizer** [comment]
You changed the order of addition properties so an operator `in` will iterate them in this order:
`local style (input-focused) -> input-base -> input`
but the correct order is:
`input-base -> input -> local style (input-focused)`
I added local style after all others

**Comment by level420** [comment]
Please do not approve currently. Still in discussion. Thank you.

**Comment by level420** [comment]
@Adolfrizer Please have again a look into the PR. I've fixed one existing problem and simplified the algorithm even more.

The "trick" is, that recursion starts with the current decorator definition, in the example `input-focused` in function `recurseDecoratorInclude` where, if an `include` property exists, it recursively calls the function with the `input` decorator which recursively calls the function with the `input-base` decorator.

As the recursion ends here, the styles get copied first over from `input-base` then from `input` and last from `input-focused`. This should be the right order, beginning from the topmost include down to the current decorator.

Additionally I've removed the unneeded clone of the `decoration` entry which seems to me to be a relic from ancient code, as not decorator entry has an attribute called `decorator`.

As only the `style` object/hash is used when creating the decorator, there is no need for `entry.styles`.

**Comment by level420** [comment]
And also note that the PR will not include the debug message anymore. This is just to illustrate that copying the styles works in the right order. 

The `name` parameter of the function `recurseDecoratorInclude` could also be used to track if we have circular references in the decorators, thus things like `dec1.include = "dec2"` and `dec2.include = "dec1"`.

**Review by Adolfrizer — APPROVED** [review]
I think it's OK. I verified it locally

**Review by cboulanger — APPROVED** [review]

**Review by hkollmann — APPROVED** [review]

**Review by johnspackman — APPROVED** [review]

**Review by cajus — APPROVED** [review]

**Review by derrell — APPROVED** [review]

**Review by oetiker — APPROVED** [review]

**Comment by johnspackman** [comment]
full approval, merging thank you @level420 !

**Comment by hkollmann** [comment]
@level420 @Adolfrizer : Can we close this issue?

**Pull request merged**